spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java

```diff
@@ -159,16 +181,152 @@ public ListenableFuture<Job> writeDataToTable(
+   * This method uses BigQuery Storage Write API to write new line delimited JSON file to the
+   * specified table. This method creates a table with the specified schema.
+   *
+   * @param tableName name of the table to write to
+   * @param jsonInputStream input stream of the json file to be written
+   * @return {@link ListenableFuture} containing the WriteApiResponse indicating completion of
+   *     operation
    */
+  @Override
+  public ListenableFuture<WriteApiResponse> writeJsonStream(
+      String tableName, InputStream jsonInputStream, Schema schema) {
+    createTable(tableName, schema); // create table if it's not already created
+    return writeJsonStream(tableName, jsonInputStream);
+  }
+
+  private void createTable(
+      String tableName, Schema schema) { // create table if it's not already created
+    TableId tableId = TableId.of(datasetName, tableName);
+    Table table = bigQuery.getTable(TableId.of(datasetName, tableName));
+    if (table == null || !table.exists()) {
+      TableDefinition tableDefinition = StandardTableDefinition.of(schema);
+      TableInfo tableInfo = TableInfo.newBuilder(tableId, tableDefinition).build();
+      bigQuery.create(tableInfo);
+    }
+  }
+
+  /**
+   * This method uses BigQuery Storage Write API to write new line delimited JSON file to the
+   * specified table. The Table should already be created as BigQuery Storage Write API doesn't
+   * create it automatically.
+   *
+   * @param tableName name of the table to write to
+   * @param jsonInputStream input stream of the json file to be written
+   * @return {@link ListenableFuture} containing the WriteApiResponse indicating completion of
+   *     operation
+   */
+  @Override
+  public ListenableFuture<WriteApiResponse> writeJsonStream(
+      String tableName, InputStream jsonInputStream) {
+
+    SettableListenableFuture<WriteApiResponse> writeApiFutureResponse =
+        new SettableListenableFuture<>();
+    CompletableFuture<Void> completableFuture =
+        CompletableFuture.runAsync(
+            () -> {
+              try {
+                WriteApiResponse apiResponse = getWriteApiResponse(tableName, jsonInputStream);
+                writeApiFutureResponse.set(apiResponse);
+              } catch (DescriptorValidationException | IOException e) {
+                writeApiFutureResponse.setException(e);
+                logger.log(Level.WARNING, String.format("Error: %s\n", e.getMessage()));
+              } catch (InterruptedException e) {
+                writeApiFutureResponse.setException(e);
+                // Restore interrupted state in case of an InterruptedException
+                Thread.currentThread().interrupt();
+              }
+            });
+
+    // register success and failure callback
+    writeApiFutureResponse.addCallback(
+        (res) -> {
+          completableFuture.cancel(true); // interrupt the completableFuture
```

Comment:
Is this "just in case"? The success callback ideally will happen after the future value is set. Completable future javadoc says that calling `cancel` is a no-op if already completed, but is there an actual scenario in which canceling an already-completed or alread-failed future is needed?

Reply:
I think calling cancel isn't needed in the `success` callback, as we would have already set the value and the `runAsync` task would have gotten completed. Is it okay if I remove the `success` callback and just keep the failure callback with the `cancel` call?

Reply:
I guess I am not sure why the callback is needed at all. Invocation of success callback means that ` writeApiFutureResponse.set(apiResponse);` got called successfully, which will lead to `completableFuture` completing. Invocation of error callback means that one of the `writeApiFutureResponse.setException(e);` calls was reached, again leading to `completableFuture`  completing correctly.
What is the state of `completableFuture` if the logic in the runnable completes successfully?

Reply:
Thanks @elefeint . I did some standalone testing for this and it appears that we do not need a success callback because of the reason you mentioned. However we might need a failure callback as the developer might call ` writeApiFutureResponse.cancel(true);` asynchronously and then we may want to interrupt the async operation which is writing the data. Please let me know if it looks good to you, we can later execute `asyncTask` on the developer's thread pool.